File: internal/apiv2/watch_together_stop_test.go
package apiv2

import (
	"context"
	"net/http"
	"strings"
	"testing"

	"github.com/Silo-Server/silo-server/internal/watchtogether"
)

type fakeRoomStop struct {
	calls, user   int
	room, profile string
	err           error
}

func (f *fakeRoomStop) StopWatchTogetherPlayback(_ context.Context, room string, user int, profile string) (watchtogether.Snapshot, string, error) {
	f.calls++
	f.user = user
	f.room = room
	f.profile = profile
	return watchtogether.Snapshot{RoomID: room, Phase: "lobby", PlaybackState: "idle", SelectionMode: "host_pick", SelectionRevision: 2, GuestControlPolicy: "host_only", AnchorUpdatedAt: "2026-01-01T00:00:00Z", SelfRole: "host", SelectedContentID: new("staged")}, "room-proof", f.err
}

func TestWatchTogetherStop(t *testing.T) {
	f := new(fakeRoomStop)
	deps := pilotDeps(nil, nil)
	deps.WatchTogetherStop = f
	h := NewHandler(deps)
	path := Prefix + "/watch-together/rooms/room/playback/stop"
	requireProblem(t, do(t, h, http.MethodPost, path, "", nil), TypeAuthenticationRequired)
	if f.calls != 0 {
		t.Fatal("unauthenticated dispatch")
	}
	rec := do(t, h, http.MethodPost, path, "", profileOwner())
	body := rec.Body.String()
	if rec.Code != 200 || !strings.Contains(body, `"phase":"lobby"`) || !strings.Contains(body, `"selected_content_id":"staged"`) || !strings.Contains(body, `"selection_revision":2`) || f.user != 1 || f.profile != "p-owner" || f.room != "room" {
		t.Fatalf("%d %s %+v", rec.Code, body, f)
	}
	for _, tc := range []struct {
		err    error
		status int
	}{{watchtogether.ErrRoomForbidden, 403}, {watchtogether.ErrRoomNotFound, 404}, {watchtogether.ErrRoomClosed, 409}} {
		f.err = tc.err
		rec = do(t, h, http.MethodPost, path, "", profileOwner())
		if rec.Code != tc.status {
			t.Fatalf("%v: %d %s", tc.err, rec.Code, rec.Body.String())
		}
	}
	deps.WatchTogetherStop = nil
	requireProblem(t, do(t, NewHandler(deps), http.MethodPost, path, "", profileOwner()), TypeDependencyUnavailable)
}

File: web/src/api/v2/watchTogetherStop.test.ts
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { setAccessToken, setProfileId, setProfileToken } from "@/api/client";
import { stopRoomPlayback } from "./watchTogetherStop";

const room = {
  room_id: "room",
  phase: "lobby",
  playback_state: "idle",
  selection_mode: "host_pick",
  selection_revision: 2,
  selected_content_id: "dune",
  code: "KX7Q2M",
  guest_control_policy: "host_only",
  is_paused: true,
  anchor_position_seconds: 0,
  anchor_updated_at: "2026-01-01T00:00:00Z",
  generation: 5,
  member_count: 2,
  host_connected: true,
  self_role: "host",
  self_can_control_transport: true,
  self_can_manage_room: true,
  self_ignore_wait: false,
};

beforeEach(() => {
  setAccessToken("login");
  setProfileId("p1");
  setProfileToken(null);
});
afterEach(() => vi.unstubAllGlobals());

describe("stopRoomPlayback", () => {
  it("posts once with the room path and returns the lobby snapshot", async () => {
    const fetch = vi.fn().mockResolvedValue(
      new Response(JSON.stringify({ room, room_access_token: "renewed" }), {
        headers: { "Content-Type": "application/json" },
      }),
    );
    vi.stubGlobal("fetch", fetch);
    const result = await stopRoomPlayback("room");
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(String(fetch.mock.calls[0]![0])).toMatch(
      /\/api\/v2\/watch-together\/rooms\/room\/playback\/stop$/,
    );
    expect(fetch.mock.calls[0]![1].method).toBe("POST");
    expect(result.room.phase).toBe("lobby");
    expect(result.room.selected_content_id).toBe("dune");
    expect(result.room_access_token).toBe("renewed");
  });

  it("aborts silently when the profile authority is replaced mid-flight", async () => {
    let resolve!: (r: Response) => void;
    vi.stubGlobal(
      "fetch",
      vi.fn().mockImplementation(() => new Promise<Response>((r) => (resolve = r))),
    );
    const pending = stopRoomPlayback("room");
    setProfileToken("replacement");
    resolve(
      new Response(JSON.stringify({ room, room_access_token: "renewed" }), {
        headers: { "Content-Type": "application/json" },
      }),
    );
    await expect(pending).rejects.toMatchObject({ name: "StaleApiRequestContextError" });
  });
});
